### PR TITLE
ARROW-11641: [CI] Use docker buildkit's inline cache to reuse build cache across different hosts

### PR DESCRIPTION
**Pull request by kszucs**
Docker buildkit supports cache inlining, which means that the cache manifest gets bundled in the produced docker image. This helps us to avoid cache misses on different hosts, greatly improving the image reusability (regarding the build speed).

This PR passes `BUILDKIT_INLINE_CACHE` build argument by default. Also adds support for using `docker buildx build` command instead of `docker build` so we can have fine grained control over the build cache. Sadly the cache manifests are not widely supported among the popular docker registries, so use the former one by default.

### CODE REVIEW

File: dev/archery/archery/cli.py [hunk 1]
```diff
@@ -767,6 +767,12 @@ def docker_compose(obj, src, dry_run):
               envvar='ARCHERY_USE_DOCKER_CLI',
               help="Use docker CLI directly for building instead of calling "
                    "docker-compose. This may help to reuse cached layers.")
+@click.option('--using-docker-buildx', default=False, is_flag=True,
+              envvar='ARCHERY_USE_DOCKER_BUILDX',
+              help="Use builx with docker CLI directly for building instead "
```
Comment (kou) [inline dev/archery/archery/cli.py:772]:
```suggestion
              help="Use buildx with docker CLI directly for building instead "
```

File: dev/archery/archery/cli.py [hunk 2]
```diff
@@ -817,6 +825,12 @@ def docker_compose_build(obj, image, *, force_pull, using_docker_cli,
...
```
Comment (kou) [inline dev/archery/archery/cli.py:830]:
```suggestion
              help="Use buildx with docker CLI directly for building instead "
```

**Review by kou — APPROVED** [review]
+1

https://github.com/apache/arrow/pull/9499/checks?check_run_id=1908923927 uses cache.

**Comment by kszucs** [comment]
Thanks! The build failures are unrelated, merging.

**Pull request closed without merging**